### ISSUE TITLE
cuModuleUnload failed

### ISSUE THREAD

**Issue by pavithrasiddesh** [arcadelab/deepdrr#45]
Hi Killen
 I am getting this cuModuleUnload failed error on setting any value to this to volume.world_from_ijk @ deepdrr.geo.point(100, 100, 100).

   return self.project(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/deepdrr/projector/projector.py", line 516, in project
    intensity, photon_prob = projector.project(proj)
  File "/usr/local/lib/python3.8/dist-packages/deepdrr/projector/projector.py", line 244, in project
    cuda.memcpy_dtoh(intensity, self.intensity_gpu)
pycuda._driver.LogicError: cuMemcpyDtoH failed: an illegal memory access was encountered
PyCUDA WARNING: a clean-up operation failed (dead context maybe?)
cuModuleUnload failed: an illegal memory access was encountered

**Comment by mathiasunberath** [comment]
without looking at code, my suspicion is that you are trying to access memory beyond the volume size because the point 100,100,100 maps beyond its extent. @benjamindkilleen we may need to catch exceptions like those though I was under the impression that the textures handle this already.

**Comment by benjamindkilleen** [comment]
@pavithrasiddesh Can you post the code snippet that produced this output, as well as the full trace?

**Comment by pavithrasiddesh** [comment]
Hi killen
 Below is the code and error that i m getting:

import deepdrr
import pydicom as dicom
import os
import numpy as np

volume
=deepdrr.Volume.from_nifti('/home/comofi/Documents/Pavithra_mn/10_recon_2_delayed.nii.gz')


camera_intrinsics = deepdrr.geo.CameraIntrinsicTransform.from_sizes(
  sensor_size=512,
  pixel_size=0.4464,
  source_to_detector_distance=780,
)


center = volume.world_from_ijk @ deepdrr.geo.point(100, 100, 100)

carm = deepdrr.CArm(
    isocenter=deepdrr.geo.point(0, 0, 0),
    isocenter_distance=780,
)

with deepdrr.Projector(volume, camera_intrinsics, carm) as projector:
  projection = projector()

on using carm as in above code i get the below  error:

(512, 512, 301)
got NifTi units: ('mm', 'unknown')
Previously, device.CArm used phi, theta as device angulation. This was
incorrect. To accomplish something similar, use device.MobileCArm instead.
Traceback (most recent call last):
  File "/home/comofi/Documents/Pavithra_mn/C-Arm
Registration/code/deepdrr_tr1.py", line 40, in <module>
    projection = projector()
  File
"/usr/local/lib/python3.8/dist-packages/deepdrr/projector/projector.py",
line 654, in __call__
    return self.project(*args, **kwargs)
  File
"/usr/local/lib/python3.8/dist-packages/deepdrr/projector/projector.py",
line 500, in project
    camera_projections = [self.carm.get_camera_projection()]
AttributeError: 'CArm' object has no attribute 'get_camera_projection'

on changing carm to mobile cam as below
carm = deepdrr.MobileCArm(isocenter=center,)

I get the following error
(512, 512, 301)
got NifTi units: ('mm', 'unknown')
PyCUDA WARNING: a clean-up operation failed (dead context maybe?)
cuArrayDestroy failed: an illegal memory access was encountered

Traceback (most recent call last):
  File "/home/comofi/Documents/Pavithra_mn/C-Arm
Registration/code/deepdrr_tr1.py", line 40, in <module>
    projection = projector()
  File
"/usr/local/lib/python3.8/dist-packages/deepdrr/projector/projector.py",
line 654, in __call__
    return self.project(*args, **kwargs)
  File
"/usr/local/lib/python3.8/dist-packages/deepdrr/projector/projector.py",
line 516, in project
    intensity, photon_prob = projector.project(proj)
  File
"/usr/local/lib/python3.8/dist-packages/deepdrr/projector/projector.py",
line 244, in project
    cuda.memcpy_dtoh(intensity, self.intensity_gpu)
pycuda._driver.LogicError: cuMemcpyDtoH failed: an illegal memory access
was encountered
PyCUDA WARNING: a clean-up operation failed (dead context maybe?)
cuModuleUnload failed: an illegal memory access was encountered

3. In another case i read the dicom files of ct volume i create nd array of
3d volume to use it in deepdrr but it shows me : (512, 512, 301)
Traceback (most recent call last):
  File "/home/comofi/Documents/Pavithra_mn/C-Arm
Registration/code/deepdrr_tr1.py", line 31, in <module>
    center = volume.world_from_ijk @ deepdrr.geo.point(100, 100, 100)
AttributeError: 'numpy.ndarray' object has no attribute 'world_from_ijk'

Basically what I have  to do is generate drr from ct volume  which has dcm
files in directory. Hence i was trying to execute sample to code understand
the working. Kindly help me in this
Regards,
Pavithra Siddesh

**Comment by benjamindkilleen** [comment]
Closing this due to lack of activity.
